Src/Base/AMReX_Algorithm.H

```diff
@@ -157,6 +157,57 @@ namespace amrex
         return hi;
     }
 
+    template<typename ItType, typename ValType>
+    AMREX_GPU_HOST_DEVICE
+    ItType upper_bound (ItType first, ItType last, const ValType& val)
+    {
+#if AMREX_DEVICE_COMPILE
+        std::ptrdiff_t count = last-first;
+        while(count>0){
+            auto it = first;
+            const auto step = count/2;
+            it += step;
+            if (!(val<*it)){
+                first = ++it;
+                count -= step + 1;
+            }
+            else{
+                count = step;
+            }
+        }
+
+        return first;
+#else
+        return std::upper_bound(first, last, val);
+#endif
+    }
+
+    template<typename ItType, typename ValType>
+    AMREX_GPU_HOST_DEVICE
+    ItType lower_bound (ItType first, ItType last, const ValType& val)
+    {
+#ifdef AMREX_DEVICE_COMPILE
+        std::ptrdiff_t count = last-first;
+        while(count>0)
+        {
+            auto it = first;
+            const auto step = count/2;
+            it += step;
+            if (!(val<=*it)){
```

Comment:
I think this should be `*it < value`, because the algorithm is only supposed to use `<`, not `<=`.

Reply:
I think that we actually need `<=` with this formulation. Otherwise with the vector  `[0,1,2,3,4,5,5,6,6,7]` passing `0` as value gives `1` instead of `0` and passing `7` gives `END` instead of `7`.

However, it could be reformulated as `if (val > *it )` . I can do that if you think that it would be better

Reply:
I am confused.  Is `if (val > *it)` mathematically equivalent to `if (*it <val)` (which I suggested) for your example?  Maybe you didn't remove `!`?

Reply:
We should only use `<`, not `>`, because a data type with `<' but without `>` works with std::lower_bound.

Reply:
Oh sorry! My fault... I misread your comment: I thought that you meant `!(val<*it)` !
You are totally right!

Reply:
It won't make difference for arithmetic types.  But the user will expect this to work too.
```
struct foo {
    int i;
    bool operator< (foo const& rhs) const {
        return i < rhs.i;
    }
    // No operator> defined
};
```

Reply:
I'll do the correction right now

Reply:
Sorry again, I've implemented your suggestion now

Reply:
No problems.  Thanks for your hard work!